Add T2B use cases

lilijap opened the pull request:
## Description

Added 3 use cases of T2B under `docs/talk2biomodels/cases/`
* IBD model (537) `Case_1.md`
* Covid model (971) `Case_2.md`
* MAP kinase model (27) `Case_3.md`

The examples include expected answers (figures, tables and text).

Mkdocs pages were generated as well.

## Fixes #54

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [ ] Added new test(s) in the ```tests``` folder
- [ ] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [X] No new tests added - md files added.
## Checklist
- [X] My code follows the style guidelines mentioned in the Code/DevOps guides
- [ ] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation (e.g. MkDocs)
- [X] My changes generate no new warnings
- [ ] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

dmccloskey commented on docs/talk2biomodels/cases/Case_2.md:
````diff
@@ -0,0 +1,44 @@
+``` Simulate the model for 50 days with an interval of 50. Plot infected cases over time. Call this model_default.```
+
+
+**Q4** Set the quarantine rate rate (`q_lockdown`) 20 times the initial value (=3.7774 e-06) and simulate the model (reproduce  figure 3C from the paper). <br>
````
```suggestion
**Q4** Set the quarantine rate rate (`q_lockdown`) 20 times the initial value (=3.7774 e-06) and simulate the model (reproduce figure 3C from the paper). <br>
```

dmccloskey commented on docs/talk2biomodels/cases/Case_1.md:
````diff
@@ -0,0 +1,156 @@
+
+**Q1:** Search for a model on BioModels database related to a search term <br>
+
+```Search for all the BioModels on Crohn's Disease```
````
At least for me, I find code snippets easier to read (and copy-n-paste) if they are chunked into code blocks like the following:

```
Search all the BioModels on Crohn's Disease
```

dmccloskey commented on docs/talk2biomodels/cases/Case_1.md:
````diff
@@ -0,0 +1,156 @@
+```Search for all the BioModels on Crohn's Disease```
+
+**A1:** 
+>| # | BioModel ID | BioModel Name | Format | Submission Date |
````
After reviewing these documents, I have to admit that it is useful to have the expected response from the LLM. We already discussed that adding the EXACT LLM response would not be the best as it will change between LLMs. Perhaps instead we could add the MINIMAL expected response. In this case, it would just be the BioModel IDs i.e., `BIOMD0000000535, BIOMD0000000536, and BIOMD0000000537`.

lilijap replied:
I agree that a pasted response to these questions would provide a lot of value for potential T2B users. I would suggest pasting the response form our latest T2B release (when the most urgent bugs are fixed), however, at the top of the markdown file I would add a disclaimer that the exact responses might change with incoming updates to the tools and upon changing a LLM. What are your thoughts on this @dmccloskey?

dmccloskey replied:
@lilijap The disclaimer is a good idea. The responses do not even have to be the exact output of the LLM i.e., just the minimal correct information that can be used to assess the LLM response is correct.

dmccloskey commented on docs/talk2biomodels/cases/Case_1.md:
````diff
@@ -0,0 +1,156 @@
+
+**Q5:** Simulate the model for 12 weeks (= 2016 hours), and give the simulation result a name. <br>
+
+```Simulate the model for 2016 hours with an interval of 2016 hours. Set the initial concentration of Dose to 200 mg. Call this result `Treatment 4wk`.```
````
A minimal expected response here could just be a table of CRP-related species concentrations over time to go with Q6. And then the minimal expected response for the plot part would be omitted since this could change based on the users requirements for the plot.

lilijap replied:
I agree that plots might change depending on user's requirements, nevertheless, I would suggest pasting exemplary figures here (with default parameters).

dmccloskey reviewed (commented):
The use cases documentation is looking really nice. As we discussed, once the features for v1/v2 of T2B are stable, a link to a video will complete the documentation.

For the video link, I would recommend YouTube. We can discuss about the details when we are ready to add the video links.

lilijap commented:
Hi @dmccloskey, 
please review the changes I made to this PR. 🥳 
In particular, please note that I added answers to prompts:
* included figures (please view the embedded figures over `mkdocs` pages.)
* included tables
* answers are hidden until the user selects them

Also I added sample questions regarding annotations of the model.

dmccloskey commented on docs/talk2biomodels/cases/Case_2.md:
```diff
@@ -0,0 +1,148 @@
+  <li>For the experiment model_high_lockdown, the value is approximately <strong>7,275.49</strong></li>
+</ul>
+</details>
+
```
Unless there is a reason for them, you can omit the trailing new lines at at the end of the file.

dmccloskey commented on docs/talk2biomodels/cases/Case_3.md:
```diff
@@ -0,0 +1,247 @@
+<summary>Answer</summary>
+
+
+<h2>Model Components of BioModel 27</h2>
```
Is there a reason that the output for this model is different than the one in use case 2 for this particular question?

lilijap replied:
I assume this is due to variability in LLMs "understanding" of these models. This one is smaller, maybe easier to comprehend?

dmccloskey commented on docs/talk2biomodels/cases/figures/C1_q11.png:
It looks like the legend is cutoff. Is this an issue with the plotly generated image or the save feature?

lilijap replied:
It's the `save` feature. Image size depends of the widow size. I fixed this image.

dmccloskey commented on mkdocs.yml:
```diff
@@ -4,8 +4,13 @@ nav:
-      - Tutorials: ./notebooks/talk2biomodels/tutorial.ipynb
+      - Introduction123: talk2biomodels/models/intro.md
+      - Case Studies:
+        - IBD: talk2biomodels/cases/Case_1.md
```
```suggestion
        - Precision medicine: talk2biomodels/cases/Case_1.md
```

dmccloskey commented on mkdocs.yml:
```diff
@@ -4,8 +4,13 @@ nav:
+      - Introduction123: talk2biomodels/models/intro.md
+      - Case Studies:
+        - IBD: talk2biomodels/cases/Case_1.md
+        - Covid-19: talk2biomodels/cases/Case_2.md
```
```suggestion
        - Epidemiology: talk2biomodels/cases/Case_2.md
```

dmccloskey commented on mkdocs.yml:
```diff
@@ -4,8 +4,13 @@ nav:
+      - Case Studies:
+        - IBD: talk2biomodels/cases/Case_1.md
+        - Covid-19: talk2biomodels/cases/Case_2.md
+        - MAPK: talk2biomodels/cases/Case_3.md
```
```suggestion
        - Emergent system properties: talk2biomodels/cases/Case_3.md
```

dmccloskey reviewed (changes requested):
The updates look really good 🤩. The use of `<details>` was a really good touch.  I only had a few small comments.

Also, are you planning the videos for this PR or a follow-up one?

gurdeep330 commented on docs/talk2biomodels/cases/Case_1.md:
````diff
@@ -0,0 +1,516 @@
+**Q5:** Simulate the model for 12 weeks (= 2016 hours), and give the simulation result a name. <br>
+
+```
+Simulate the model for 2016 hours with an interval of 2016 hours. Set the initial concentration of Dose to 200 mg. Call this result `Treatment 4wk`.
````
Using a time interval of 2016 will generate a table with 2016 rows. While this works well for the `simulate_model` tool, it is too large (context window) for the `ask_question` tool. The `ask_question` tool relies on an inbuilt dataframe agent to query simulation results effectively. It performs well with tables up to **500 rows**, with a maximum limit of **1000 rows**. Beyond this, performance degrades, query execution time increases, and in some cases, the agent may fail to generate an answer.  

I have documented the limitation in the **How-to Guide**, and will implement an alternative in the future. I recommend using a time interval of **less than 500** to ensure optimal performance.

lilijap replied:
Thank you for pointing this out. I reduced the step size for these simulations accordingly.

gurdeep330 commented on docs/talk2biomodels/cases/Case_1.md:
````diff
@@ -0,0 +1,516 @@
+**Q8:** Administrate the drug treatment every two weeks by activating the `DoseQ2W` and deactivating `Dose` parameters. <br>  
+
+```
+Simulate the model again for 2016 hours with an interval of 2016 hours. Set the initial concentration of `DoseQ2W` to 200 and `Dose` to 0. Call this result `Treatment 2wk`.
````
Ditto as above.

gurdeep330 commented on docs/talk2biomodels/cases/Case_1.md:
````diff
@@ -0,0 +1,516 @@
+**Q4.4:** Interrogate specific species in the model <br>
+
+```
+what is R in this model?
````
This is also a great place to highlight the use case of the `query_article` tool.

_Query the uploaded article to suggest a possible description of the parameter R in the model._

lilijap replied:
Thank you for this suggestion! I added two RAG-related questions to this example.

gurdeep330 reviewed (changes requested):
Hi @lilijap 

Thanks for putting together the use cases. Nice work 💯
I have some minor comments/suggestions. Let me know your thoughts.

I agree with @dmccloskey that adding a video URL would enhance the content. I suggest including a separate video for each case, placing it at the end of its respective section. You could create one long video and then trim it into three segments. Clipchamp in Windows is a good tool for trimming and adding subtitles to the videos.

lilijap commented:
Thank you @dmccloskey and @gurdeep330 for your valuable comments and suggestions! 🙏 
Regarding your comments: I would like to address following changes **in this PR**:
- adjust prompts to **smaller number of time-steps** (see @gurdeep330 comments on the context window) to optimize the reasoning over simulation results
- add an example of **RAG** over a paper to the model 537

In the **next PR(s)**:
- adding **videos** accompanying the use cases. Currently questions and examples have not reached a "steady state" and are likely to change in the course of this PR. I would like to produce a video using stable use cases.
- adding a use-case for an **uploaded model**

Please let me know, if you disagree! 😃

lilijap commented:
Dear @dmccloskey and @gurdeep330, I would merge the branch, as soon as I have your formal & final approval of this PR. Thanks!

dmccloskey reviewed (approved):
👍

gurdeep330 reviewed (approved):
Thanks!
